osh/sh_expr_eval.py

```diff
@@ -327,6 +328,12 @@ def _EvalLhsAndLookupArith(self, node):
     lval = self.EvalArithLhs(node, runtime.NO_SPID)
     val = OldValue(lval, self.mem, self.exec_opts)
 
+    # BASH_LINENO, etc.
+    if val.tag_() in (value_e.MaybeStrArray, value_e.AssocArray) and lval.tag_() == lvalue_e.Named:
+      named_lval = cast(lvalue__Named, lval)
+      if word_eval.CheckCompatArray(named_lval.name, self.exec_opts):
+        val = word_eval.ResolveCompatArray(val)
```

Comment:
Would this part be made obsolete if we made BASH_LINENO etc. immutable?

I guess that would make Oil technically incompatible, but I've been thinking about that.  Also it makes sense to tighten up things like `PATH=(a b c)` and `HOME=(a b c)` (they shouldn't be arrays)

This doesn't have to be done in this commit; I'm just curious / brainstorming the semantics

Reply:
As you have already written in another comment, this change also affects the other mutable variables when `shopt -s compat_array` is set.